src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -31,6 +32,27 @@ impl ScalarFunc {
         let res = t.date_format(format_mask_str)?;
         Ok(Some(Cow::Owned(res.into_bytes())))
     }
+
+    #[inline]
+    pub fn date<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        if t.invalid_zero() {
+            return Err(box_err!("Incorrect datetime value: '{}'", t));
+        }
+        if t.is_zero() {
+            return Err(box_err!("Datetime value is zero timestamp: '{}'", t));
```

Comment:
I think we need to return an error with the same error code as tidb did, you can define it similar to https://github.com/tikv/tikv/blob/master/src/coprocessor/codec/error.rs#L69
Also you need to handle the error with ctx like tidb did in `handleInvalidTimeError`